Duplicati/Library/Utility/FilterGroups.cs

```diff
@@ -0,0 +1,702 @@
+
+            if (Utility.IsClientOSX)
+                osFilters = CreateOSXFilters(group);
+            else if (Utility.IsClientLinux)
+                osFilters = CreateLinuxFilters(group);
+            else if (Utility.IsClientWindows)
+                osFilters = CreateWindowsFilters(group);
+            else
+                throw new ArgumentException("Unknown operating system?");
+
+            return
+                FilterPrefixMatches(
+                    CreateCommonFilters(group)
+                        .Concat(osFilters)
+                        .Where(x => !string.IsNullOrWhiteSpace(x))
+                    )
+                    .Distinct(Utility.ClientFilenameStringComparer);
+        }
+
+        /// <summary>
+        /// Filters all items that have a prefix
+        /// </summary>
+        /// <returns>The prefix matches.</returns>
+        /// <param name="filters">Filters.</param>
+        private static IEnumerable<string> FilterPrefixMatches(IEnumerable<string> filters)
+        {
+            // Any paths that start with "?:" or "*:" can refer to a path on any drive.
+            // If we see any of them, then we remember them, so that if we then see a path that starts with an absolute drive letter
+            // ("C:" or "D:") we can remove it if the path matches.
+            HashSet<string> wildcardRootWindowsPaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            string prev = null;
+            foreach (var n in filters.OrderBy(x => x, Utility.ClientFilenameStringComparer))
+            {
+                if (n.StartsWith("?:", StringComparison.Ordinal) || n.StartsWith("*:", StringComparison.Ordinal))
+                {
+                    wildcardRootWindowsPaths.Add(n.Substring(2));
+                }
+
+                if (n.Length > 2 && char.IsLetter(n[0]) && n[1] == ':' && wildcardRootWindowsPaths.Contains(n.Substring(2)))
+                    continue;
+                else if (prev != null && prev.EndsWith(System.IO.Path.DirectorySeparatorChar.ToString(), StringComparison.Ordinal) && n.StartsWith(prev, Utility.ClientFilenameStringComparision))
+                    continue;
+                else
+                    yield return prev = n;
+            }
+        }
+
+        /// <summary>
+        /// Creates common filters
+        /// </summary>
+        /// <param name="group">The groups to create the filters for</param>
+        /// <returns>Common filters</returns>
+        private static IEnumerable<string> CreateCommonFilters(FilterGroup group)
+        {
+            if (group.HasFlag(FilterGroup.CacheFiles))
+            {
+                // TODO: The control_dir_v2 might be under a different path for OEM branded instances.
+                // However, the AppName is loaded and controlled by the AutoUpdater assembly, which we can't reference here without an ugly circular dependency or dependency injection.
+                // What is the best way to solve this?

```

Comment:
Lets leave it for now. We can do it by reflection, loading the assembly at runtime, and extracting the path but I don't want to delay merging it just to support that.

Reply:
Opened #3175 for this.